Doc/library/pathlib.rst

```diff
@@ -1633,6 +1633,34 @@ Copying, renaming and deleting
       Added return value, return the new :class:`!Path` instance.
 
 
+.. method:: Path.delete(ignore_errors=False, on_error=None)
+
+   Recursively remove this file or directory tree.
+
+   If *ignore_errors* is true, errors resulting from failed removals will be
+   ignored. If *ignore_errors* is false or omitted, and a function is given to
+   *on_error*, it will be called each time an exception is raised. If neither
+   *ignore_errors* nor *on_error* are supplied, exceptions are propagated to
+   the caller.
+
+   If the optional argument *on_error* is specified, it should be a callable;
+   it will be called with one argument of type :exc:`OSError`. The
+   callable can handle the error to continue the deletion process or re-raise
+   it to stop. Note that the filename is available as the
+   :attr:`~OSError.filename` attribute of the exception object.
+
+   .. note::
+
+      On platforms that support the necessary fd-based functions, a symlink
+      attack-resistant version of :meth:`~Path.delete` is used to delete
+      directory trees. On other platforms, the :func:`~Path.delete`
+      implementation is susceptible to a symlink attack: given proper timing
+      and circumstances, attackers can manipulate symlinks on the filesystem
+      to delete files they would not be able to access otherwise.
```

Comment:
This will need re-wrapping if accepted. 

Can we outline what platforms are/aren't at risk? This is quite vague currently!
```suggestion
      This function can be susceptible to a 'symlink attack':
      given proper timing and circumstances, 
      attackers can manipulate symlinks on the filesystem
      to delete files they would not be able to access otherwise.
      
      On platforms that support the necessary file descriptor-based functions, 
      a symlink attack-resistant version of :meth:`~Path.delete` is used to delete
      directory trees. 
      On other platforms, the :func:`~Path.delete`
      implementation is susceptible to a this form of attack.
```

Reply:
> Can we outline what platforms are/aren't at risk? This is quite vague currently!

It's complicated:

```python
({os.open, os.stat, os.unlink, os.rmdir} <= os.supports_dir_fd and
 os.listdir in os.supports_fd and os.stat in os.supports_follow_symlinks))
```

`shutil` provides an [`rmtree.avoids_symlink_ataccks`](https://docs.python.org/3/library/shutil.html#shutil.rmtree.avoids_symlink_attacks) attribute. We could do the same thing here, but we'd need to `import shutil` at a module level rather than inline. Personally I think it's worthwhile - what do you reckon?

Reply:
Ouch! Highly cryptic.

I agree that ``Path.rmtree.avoids_symlink_attacks`` is valuable to have, not least as it means that all ``shutil`` uses can be directly translated. The only imports it would add to ``pathlib`` are ``zlib`` / ``bz2`` / ``lzma``, which take 60, 1600, and 900 microseconds on my computer, and `shutil` itself, which takes 1150, for a total of ~3.7ms extra import time.

I am concerned though that ``shutil.rmtree.avoids_symlink_attacks is False`` on Windows, though not much we can do about that here!

A

Reply:
Thanks! I've added `Path.rmtree.avoids_symlink_attacks`